### PR TITLE
dj0: expanded blob functionality

### LINKED ISSUE #244: Serializing non `ndarray` objects
Currently BLOB field on Python side only accepts inserting `ndarrays` of simple data types. We should provide broader serialization support on the Python side.

This, however, brings up an issue of how _compatible_ we should keep Python serialization with DataJoint matlab. Namely, should we allow Python to serialize objects that cannot be deserialized in Matlab side? Equivalently, should we allow Python to serialize via serialization scheme other than `mYm`?

### LINKED ISSUE #392: Serializing dict as blob assumes string keys
It seems dict serialization was added last year in #248 by @eywalker . I'm trying to add a dict as a mediumblob to our database. The dict has integer keys, not string keys, and they map to tuples. It looks something like this:
```
{1:(0, 100), 2:(0, 200)}
```
Currently, datajoint assumes that keys are always strings:
https://github.com/datajoint/datajoint-python/blob/master/datajoint/blob.py#L286

I tried changing `pack_string(k)` to `pack_obj(v)`, which works fine while pack, but later I get a byte count assertion error when unpacking the blob here:
https://github.com/datajoint/datajoint-python/blob/master/datajoint/blob.py#L111

Any ideas?

### LINKED ISSUE #427: Expand blob serialization functionality
This issue pulls together issues related to blob serialization to propose a single solution to address them.

This includes issues #392 and #244. 

The goal is to allow storing arbitrary Python objects if MATLAB compatibility is not important.  Clarify what objects are portable between MATLAB and Python and allow configuring alternative serialization and compression protocols.

### LINKED ISSUE #475: subfolders for external storage
To avoid millions of files in one folder, break up external storage files into subfolders.

1. Add subfolder scheme to the external storage configuration in `dj.config`.  The structure may be specified as a tuple, e.g. (2,2,2), which indicates three levels of subfolders, each taking the next 2 characters of the hash and leaving the rest for the filename.  The default scheme is (), i.e. no subfolders.

2.  When creating a new external storage folder, create the file `djexternal.json` and save the subfolder scheme.  

3. When using an existing external storage folder, find `djexternal.json` and read the subfolder scheme from it, ignoring the one specified in `dj.config`.

4. Break up the hash filename according to the subfolder scheme for both reading and writing.

### LINKED ISSUE #520: Python Blob cannot store list of Strings
When storing a list of data inside a blob, string datatypes are not functional in Python. 

When attempting to store a string list into a blob an error like the following will be thrown:
```
~/code/Python_3/lib/python3.7/site-packages/datajoint/blob.py in pack_array(array)
    252         array, imaginary = np.real(array), np.imag(array)
    253 
--> 254     type_number = rev_class_id[array.dtype]
    255 
    256     if dtype_list[type_number] is None:

KeyError: dtype('<U1')
```

### LINKED ISSUE #572: np.float32 not supported for longblobs yet
I was attempting to shove a numpy array of FP32 values and datajoint complained that it is not supported yet.

Error message:
packing object of type <class 'numpy.float32'> currently not supported!

I got around it by converting it to np.float which is equivalent to np.float64, but that will be taking up double the space.

Is this a feature that is going be implemented later? Also will there be support for other datatypes such as np.float16?

### LINKED ISSUE #594: uuid string insert/restrict support
Since uuid's have a clear string to value mapping, using string versions may be more user friendly in some cases where it would be possible to support them.

For the moment, this is:

- record insert
- restrict via dictionary

### PR DESCRIPTION
**Pull request by dimitri-yatsenko**
Blob now supports serialization of simple Python data types: `dict`, `tuple`, `list`, `set`, `string`, and `bytes`, `np.array` of `dtype('O')`, numeric scalars, `bool`, matlab cell arrays, and matlab struct arrays, `datetime.datetime`, `datetime.date`, `datetime.time`.

* Fix #572
* Fix #520 
* Fix #427
* Fix #392 
* Fix #244

* Fix #594 -- support string values for UUID attributes.

### CODE REVIEW

**Comment by guzman-raphael** [comment]
Following my review with @eywalker, the following questions came up:

1. Why is the record array representation denoted by `F` as opposed to `f`? Thought this was agreed to be lower case.

https://github.com/dimitri-yatsenko/datajoint-python/blob/25965894484a82de68a6d4dde46e64291214e997/datajoint/blob.py#L121

2. Was there not a lighter way to pack a decimal e.g. 2 integers? Curious why string was utilized.

https://github.com/dimitri-yatsenko/datajoint-python/blob/25965894484a82de68a6d4dde46e64291214e997/datajoint/blob.py#L253-L256

3. Using the default `encode` method utilizes UTF8. Are we confident that all python string values can be represented in UTF8?

https://github.com/dimitri-yatsenko/datajoint-python/blob/25965894484a82de68a6d4dde46e64291214e997/datajoint/blob.py#L261-L264

File: datajoint/blob.py [hunk 1]
```diff
@@ -34,286 +37,384 @@
...
+                b"".join(self.pack_recarray(array[f]) if array[f].dtype.fields else self.pack_array(array[f])
+                         for f in array.dtype.names))
+
+    def read_sparse_array(self):
+        raise DataJointError('datajoint-python does not yet support sparse arrays. Issue (#590)')
+
+    def read_decimal(self):
+        return Decimal(self.read_string())
+
+    @staticmethod
```
Comment (guzman-raphael) [inline datajoint/blob.py]: Was there not a lighter way to pack a decimal e.g. 2 integers? Curious why string was utilized.

File: datajoint/blob.py [hunk 2]
```diff
@@ -34,286 +37,384 @@
...
+    def pack_decimal(d):
+        s = str(d)
+        return b"d" + len_u64(s) + s.encode()
+
+    def read_string(self):
+        return self.read_binary(self.read_value()).decode()
+
+    @staticmethod
```
Comment (guzman-raphael) [inline datajoint/blob.py]: Using the default encode method utilizes UTF8. Are we confident that all python string values can be represented in UTF8?

File: datajoint/blob.py [hunk 3]
```diff
@@ -34,286 +37,384 @@
...
+    def read_sparse_array(self):
+        raise DataJointError('datajoint-python does not yet support sparse arrays. Issue (#590)')
+
+    def read_decimal(self):
+        return Decimal(self.read_string())
+
+    @staticmethod
+    def pack_decimal(d):
+        s = str(d)
+        return b"d" + len_u64(s) + s.encode()
```
Comment (guzman-raphael) [inline datajoint/blob.py]: Was there not a lighter way to pack a decimal e.g. 2 integers? Curious why string was utilized.

File: datajoint/blob.py [hunk 4]
```diff
@@ -34,286 +37,384 @@
...
+
+    def read_string(self):
+        return self.read_binary(self.read_value()).decode()
+
+    @staticmethod
+    def pack_string(s):
+        blob = s.encode()
+        return b"\5" + len_u64(blob) + blob
```
Comment (guzman-raphael) [inline datajoint/blob.py]: Using the default encode method utilizes UTF8. Are we confident that all python string values can be represented in UTF8?

File: datajoint/blob.py (hunk 3 above)
Comment (dimitri-yatsenko) [reply datajoint/blob.py]: I do not know if two integers are enough. Could not find documentation.

File: datajoint/blob.py [hunk 5]
```diff
@@ -34,286 +37,384 @@
...
+                "C": self.read_cell_array,    # matlab cell array
+                # Python-native
+                "\xFF": self.read_none,      # None
+                "\1": self.read_tuple,     # a Sequence
+                "\2": self.read_list,      # a MutableSequence
+                "\3": self.read_set,       # a Set
+                "\4": self.read_dict,      # a Mapping
+                "\5": self.read_string,    # a UTF8-encoded string
+                "\6": self.read_bytes,     # a ByteString
+                "F": self.read_recarray,   # numpy array with fields, including recarrays
```
[the remark this answers is not in the document]
Comment (dimitri-yatsenko) [reply datajoint/blob.py]: Just following the pattern that all arrays are uppercase.

File: datajoint/blob.py (hunk 4 above)
Comment (dimitri-yatsenko) [reply datajoint/blob.py]: What is the alternative. UTF8 is dominant and continues to grow.

File: datajoint/blob.py [hunk 6]
```diff
@@ -34,286 +37,384 @@
...
+                         for f in array.dtype.names))
+
+    def read_sparse_array(self):
+        raise DataJointError('datajoint-python does not yet support sparse arrays. Issue (#590)')
+
+    def read_decimal(self):
+        return Decimal(self.read_string())
+
+    @staticmethod
+    def pack_decimal(d):
```
[the remark this answers is not in the document]
Comment (dimitri-yatsenko) [reply datajoint/blob.py]: I have looked around and could not find the optimal robust representation. string seems common.

File: datajoint/blob.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dimitri-yatsenko) [reply datajoint/blob.py]: I followed the pattern that arrays are capitalized.

**Comment by chrisroat** [comment]
Do you know if this would allow storage of general pandas DataFrames as blobs?

**Comment by ixcat** [comment]
hi @chrisroat , just happened to be reviewing this issue while working elsewhere - pandas inserts should be functional (as well as fetch(... format='frame') - see also: `https://github.com/datajoint/datajoint-python/pull/776`

**Pull request merged**